docs/core/testing/unit-testing-mstest-writing-tests-testcontext.md

```diff
@@ -0,0 +1,183 @@
+- `TestRunResultsDirectory` – the directory where the test results are stored. Typically a subdirectory of the `ResultsDirectory`.
+- `TestResultsDirectory` – the directory where the test results are stored. Typically a subdirectory of the `ResultsDirectory`.
+
+In MSTest 3.7 and later, the `TestContext` class also provides new properties helpful for `TestInitialize` and `TestCleanup` methods:
+
+- `TestData` – the data that will be provided to the parameterized test method or `null` if the test is not parameterized.
+- `TestDisplayName` - the display name of the test method.
+- `TestException` - the exception thrown by either the test method or test initialize or `null` if the test method did not throw an exception.
+
+### Data-driven tests
+
+`TestContext` is essential for data-driven testing in MSTest. It enables you to retrieve and set data for each iteration in a data-driven test, using properties like `DataRow` and `DataConnection` (for datasource based tests).
+
+Assuming the following CSV file `TestData.csv`:
+
+```csv
+Number,Name
+1,TestValue1
+2,TestValue2
+3,TestValue3
+```
+
+You can use the `DataSource` attribute to read the data from the CSV file:
+
+```csharp
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using System;
+
+namespace YourNamespace
+{
+    [TestClass]
+    public class CsvDataDrivenTest
+    {
+        public TestContext TestContext { get; set; }
+
+        [TestMethod]
+        [DataSource(
+            "Microsoft.VisualStudio.TestTools.DataSource.CSV", 
+            "|DataDirectory|\\TestData.csv", 
+            "TestData#csv", 
+            DataAccessMethod.Sequential)]
+        public void TestWithCsvDataSource()
+        {
+            // Access data from the current row
+            int number = Convert.ToInt32(TestContext.DataRow["Number"]);
+            string name = TestContext.DataRow["Name"].ToString();
+
+            Console.WriteLine($"Number: {number}, Name: {name}");
+
+            // Example assertions or logic
+            Assert.IsTrue(number > 0);
+            Assert.IsFalse(string.IsNullOrEmpty(name));
+        }
+    }
+}
+```
+
+In MSTest 3.7 and later, the property `TestData` can be used to access the data for the current test during `TestInitialize` and `TestCleanup` methods.
+
+### Storing and Retrieving Runtime Data
+
+You can use `TestContext.Properties` to store custom key-value pairs that can be accessed across different methods in the same test session.
+
+```csharp
+TestContext.Properties["MyKey"] = "MyValue";
+string value = TestContext.Properties["MyKey"]?.ToString();
+```
+
+### Associating data to a test
+
+The `TestContext.AddResultFile` method allows you to add a file to the test results, making it available for review in the test output. This can be useful if you generate files during your test (e.g., log files, screenshots, or data files) that you want to attach to the test results.
```

Comment:
```suggestion
The `TestContext.AddResultFile` method allows you to add a file to the test results, making it available for review in the test output. This can be useful if you generate files during your test (for example, log files, screenshots, or data files) that you want to attach to the test results.
```